url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+          System.err.println("Java!! You ain't serious, are you? Anyway, couldn't put message on sendQ in Controller");
+          e.printStackTrace();
+        }
+    }
+
+
+
+	@SuppressWarnings("unchecked")
+    public void sendJoin(InetSocketAddress bootstrap_node,
+                         int nodeHash,
+                         int bootHash){
+        nodeID = nodeHash;
+        bootstrapNode = bootstrap_node;
+        if(nodeHash == bootHash) return;
+
+        JSONObject sendJSON = new JSONObject();
+        sendJSON.put("type",MessageType.JOINING_NETWORK_SIMPLIFIED.name());
+        sendJSON.put("node_id",Integer.toString(nodeHash));
+        sendJSON.put("target_id",Integer.toString(bootHash));
+        sendJSON.put("ip_address",NetworkProperties.nAddress.toString());
+        try {
+            sendQueue.put(jsonToDatagram(sendJSON,bootstrap_node));
+        } catch (InterruptedException e) {
+            System.err.println("Could not put Datagram on sendQueue in Controller");
+            e.printStackTrace();
+        }
+
+    }
+	@SuppressWarnings("unchecked")
+	public void leaveNetwork(long network_id) {
+        //tell of me contact I am leaving
+		for (Map.Entry<Integer, Contact> entry : routingTable.entrySet()){
+			JSONObject message = new JSONObject();
+			message.put("type", "LEAVE_NETWORK");
+			message.put("node_id", String.valueOf(nodeID));
+			InetSocketAddress address = new InetSocketAddress(entry.getValue().getIp(),NetworkProperties.nPort); 
+			try {
+	            sendQueue.put(jsonToDatagram(message,address));
+	        } catch (InterruptedException e) {
+	            System.err.println("Could not put Datagram on sendQueue in Controller");
+	            e.printStackTrace();
+	        }
+		}	
+		
+	}
+	@SuppressWarnings("unchecked")
+	public void sendIndex(String url, String[] unique_words) {// this url has these words. send an index message for each word
+		JSONObject template = new JSONObject();
+		template.put("type", "INDEX");
+		for(String word: unique_words){
+			JSONObject message = (JSONObject) template.clone();
+			int wordHash = NetworkProperties.hashCode(url);
+			if(wordHash == nodeID){//if word hash matches my id then I should index
+				SearchResponse index = new SearchResponse();
+				index.word = word;
+				index.urls.add(new Url(url,1));
+				indexer.index(index);
+			}else{ // else send to closest node. Might even match exactly
+				Contact closestNode = getClosestNode(wordHash);
+				if(closestNode != null){
+					JSONArray links = new JSONArray();
+					links.add(word);
+					message.put("keyword", word);
+					message.put("link", links);
+					message.put("sender_id", String.valueOf(nodeID));
+					message.put("target_id", String.valueOf(wordHash));
+					InetSocketAddress address = new InetSocketAddress(closestNode.getIp(),NetworkProperties.nPort); 
+					try {
+			            sendQueue.put(jsonToDatagram(message,address));
+			        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)